File: src/main/java/com/fengsheng/handler/add_robot_tos.java
package com.fengsheng.handler;

import com.fengsheng.*;
import com.fengsheng.protos.Errcode;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

public class add_robot_tos extends AbstractProtoHandler<Fengsheng.add_robot_tos> {
    private static final Logger log = Logger.getLogger(add_robot_tos.class);

    @Override
    protected void handle0(HumanPlayer player, Fengsheng.add_robot_tos pb) {
        synchronized (Game.class) {
            if (player.getGame().isStarted()) {
                log.error("room is already full");
                return;
            }
            if (player.getGame().getPlayers().length > 5) {
                player.send(Errcode.error_code_toc.newBuilder().setCode(Errcode.error_code.robot_not_allowed).build());
                return;
            }
            Statistics.PlayerGameCount count = Statistics.getInstance().getPlayerGameCount(player.getPlayerName());
            if (count == null || count.winCount() <= 0) {
                long now = System.currentTimeMillis();
                long startTrialTime = Statistics.getInstance().getTrialStartTime(player.getDevice());
                if (startTrialTime != 0 && now - 5 * 24 * 3600 * 1000 >= startTrialTime) {
                    player.send(Errcode.error_code_toc.newBuilder().setCode(Errcode.error_code.robot_not_allowed).build());
                    return;
                }
                Statistics.getInstance().setTrialStartTime(player.getDevice(), now);
            }
            Player robotPlayer = new RobotPlayer();
            robotPlayer.setPlayerName(Player.randPlayerName());
            robotPlayer.setGame(player.getGame());
            robotPlayer.getGame().onPlayerJoinRoom(robotPlayer, Statistics.getInstance().getTotalPlayerGameCount());
        }
    }
}

File: src/main/kotlin/handler/AddRobotTos.kt
package com.fengsheng.handler

import com.fengsheng.*
import com.fengsheng.protos.Fengsheng

class AddRobotTos : AbstractProtoHandler<Fengsheng.add_robot_tos>() {
    override fun handle0(r: HumanPlayer, pb: Fengsheng.add_robot_tos) {
        if (r.game!!.isStarted) {
            return
        }
        val emptyPosition = r.game!!.players.count { it == null }
        if (emptyPosition == 0) {
            return
        }
        if (!Config.IsGmEnable) {
//            val score = Statistics.getScore(r.playerName) ?: 0
//            if (score <= 0) {
//                val now = System.currentTimeMillis()
//                val startTrialTime = Statistics.getTrialStartTime(r.playerName)
//                if (startTrialTime == 0L) {
//                    Statistics.setTrialStartTime(r.playerName, now)
//                } else if (now - 3 * 24 * 3600 * 1000 >= startTrialTime) {
//                    r.sendErrorMessage("您已被禁止添加机器人，多参与群内活动即可解锁")
//                    return
//                }
//            }
            val humanCount = r.game!!.players.count { it is HumanPlayer }
            if (humanCount <= 1 && emptyPosition == 1) {
                if (Statistics.getEnergy(r.playerName) <= 0) {
                    r.sendErrorMessage("""你的精力不足，不能进行人机局。你可以在群里输入"精力系统"了解。""")
                    return
                }
                if (Game.gameCache.count { (_, v) -> // 未开始或有空位的房间（含本房间）大于1，则禁止开局
                        !v.isStarted || v.players.any { it !is HumanPlayer }
                    } > 1) {
                    r.sendErrorMessage("""有其他玩家正在进行游戏，请等待他们一起吧！""")
                    return
                }
            }
        }
        val robotPlayer = RobotPlayer()
        robotPlayer.playerName = Player.randPlayerName(r.game!!)
        robotPlayer.game = r.game
        robotPlayer.game!!.onPlayerJoinRoom(robotPlayer, Statistics.totalPlayerGameCount.random())
    }
}
